File: getmediatoken.rb
require 'base64'
require 'json'
require 'pathname'
require 'stringio'
require 'openssl'

def main(args)
  version = 1
  enddt = -1.0
  mxc = args[1]

  message = [enddt].pack("G") + mxc.encode("UTF-8")

  keyfile = Pathname.new(args[0])
  keydata = JSON.parse(keyfile.read)
  key = Base64.urlsafe_decode64(keydata["k"] + '=')
  signature = OpenSSL::HMAC.digest("SHA512", key, message)

  token = [version].pack("C") + signature + message
  puts Base64.urlsafe_encode64(token).gsub("=", "")
end

if __FILE__ == $0
  main(ARGV[0..-1])
end

File: getmediatoken.py
import base64, hashlib, hmac, json, pathlib, struct, sys
from typing import Any, Sequence


def main(args: Sequence[str]) -> None:
	version: int = 1
	enddt: float = -1.0
	mxc: str = args[1]
	message: bytes = struct.pack(">d", enddt) + mxc.encode("UTF-8")
	
	keyfile: pathlib.Path = pathlib.Path(args[0])
	keydata: Any = json.loads(keyfile.read_bytes())
	key: bytes = base64.urlsafe_b64decode(keydata["k"] + '=')
	signature: bytes = hmac.digest(key, message, hashlib.sha512)
	
	token: bytes = struct.pack(">B", version) + signature + message
	print(base64.urlsafe_b64encode(token).decode("UTF-8").replace("=", ""))


if __name__ == "__main__":
	main(sys.argv[1 : ])
